File: VB.NET/LizardKingStudios.Services/LizardKingStudios.Services.SpellcheckEngine/SpellcheckEngine.vb
Imports System
Imports System.Reflection
Imports Microsoft.Office.Interop.Word

Public Class SpellcheckEngine

    Private _wordSpellChecker As Application = New Application()
    Private _maxSuggestionsCount As Integer = 5

    Public Sub New()

        Dim optionalValue As Object = Missing.Value

        _wordSpellChecker.Application.Documents.Add(optionalValue, optionalValue, optionalValue, optionalValue)

    End Sub

    Public Property MaxSuggestionsCount() As Integer
        Get
            Return _maxSuggestionsCount
        End Get
        Set(ByVal value As Integer)
            _maxSuggestionsCount = value
        End Set
    End Property

    Public Function Execute(ByVal checkValue As String) As String

        Dim suggestedValue As String = String.Empty

        If (checkValue.Length <> 0) Then

            Dim optionalValue As Object = Missing.Value
            Dim ignoreUppercase As Object = True
            Dim suggestionMode As Object = Missing.Value

            Dim theSuggestions As SpellingSuggestions = _wordSpellChecker.GetSpellingSuggestions(checkValue, optionalValue, _
                optionalValue, optionalValue, optionalValue, _
                optionalValue, optionalValue, optionalValue, optionalValue, _
                optionalValue, optionalValue, optionalValue, optionalValue, _
                optionalValue)

            ' if we get more than (parameter specified) suggestions, don't make
            ' the change.
            If (theSuggestions.Count > 0 And theSuggestions.Count <= _maxSuggestionsCount) Then

                Try
                    If (theSuggestions.Count > 1) Then
                        Dim theUserConfirmWindow As userConfirm = New userConfirm()
                        theUserConfirmWindow.suggestionsList = theSuggestions
                        theUserConfirmWindow.ShowDialog()
                        suggestedValue = theUserConfirmWindow.selectedValue
                    Else

                        Dim theSuggestion As SpellingSuggestion = theSuggestions(1)
                        suggestedValue = theSuggestion.Name

                    End If
                Catch
                    ' lets swallow it - implement error handling here!
                End Try

            End If

        End If

        Return suggestedValue

    End Function

End Class

File: CS/LizardKingStudios.Services/LizardKingStudios.Services.SpellCheck/SpellcheckEngine.cs
using System;
using System.Reflection;
using Microsoft.Office.Interop.Word;

namespace LizardKingStudios.Services.SpellCheck
{
    public class SpellcheckEngine
    {
        private readonly Application _wordSpellChecker = new Application();
        private int _maxSuggestionsCount = 5;

        public SpellcheckEngine()
        {
            object optional = Missing.Value;

            _wordSpellChecker.Application.Documents.Add(ref optional, ref optional, ref optional, ref optional);
        }

        public int MaxSuggestionsCount
        {
            get
            {
                return _maxSuggestionsCount;
            }
            set
            {
                _maxSuggestionsCount = value;
                
            }
        }
        
        public string Execute(string checkValue)
        {
            string suggestedValue = String.Empty;

            if (checkValue.Length!=0)
            {
                object optional = Missing.Value;
                object ignoreUppercase = true;
                object suggestionMode = Missing.Value;

                SpellingSuggestions theSuggestions = _wordSpellChecker.GetSpellingSuggestions(checkValue, ref optional,
                    ref optional, ref optional, ref optional, 
                    ref optional, ref optional, ref optional, ref optional, 
                    ref optional, ref optional, ref optional, ref optional, 
                    ref optional);

                // if we get more than (parameter specified) suggestions, don't make
                // the change.
                if (theSuggestions.Count > 0 && theSuggestions.Count <= _maxSuggestionsCount)
                {
                    try
                    {
                        if (theSuggestions.Count > 1)
                        {
                            userConfirm theUserConfirmWindow = new userConfirm();
                            theUserConfirmWindow.suggestionsList = theSuggestions;
                            theUserConfirmWindow.ShowDialog();
                            suggestedValue = theUserConfirmWindow.selectedValue; 
                        }
                        else
                        {
                            SpellingSuggestion theSuggestion = theSuggestions[1];
                            suggestedValue = theSuggestion.Name;
                        }

                        
                    }
                    catch
                    {
                        // lets swallow it - implement error handling here!
                    }
                }
            }

            return suggestedValue;

        }
    }
}
